iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/node/WALNode.java

```diff
@@ -255,91 +284,108 @@ public void run() {
+            identifier,
+            firstValidVersionId);
         return;
       }
-      logger.debug(
-          "Effective information ratio {} (active memTables cost is {}, flushed memTables cost is {}) of wal node-{} is below wal min effective info ratio {}, some memTables will be snapshot or flushed.",
-          effectiveInfoRatio,
-          costOfActiveMemTables,
-          costOfFlushedMemTables,
-          identifier,
-          config.getWalMinEffectiveInfoRatio());
-      boolean isSuccess = snapshotOrFlushMemTable();
-      if (isSuccess && recursionTime < MAX_RECURSION_TIME) {
-        // wal is used to search, cannot optimize files deletion
-        if (safelyDeletedSearchIndex != DEFAULT_SAFELY_DELETED_SEARCH_INDEX) {
-          return;
+
+      buffer.subtractDiskUsage(deleteFileSize);
+      buffer.subtractFileNum(successfullyDeleted.size());
+      for (int i = 0; i < successfullyDeleted.size(); i++) {
+        long versionId = WALFileUtils.parseVersionId(successfullyDeleted.get(i));
+        Long memTableRamCostSum = walFileVersionId2MemTablesTotalCost.remove(versionId);
+        if (memTableRamCostSum != null) {
+          totalCostOfFlushedMemTables.addAndGet(-memTableRamCostSum);
         }
-        recursionTime++;
-        run();
+      }
+
+      if (!pinnedMemTableIds.isEmpty()
+          || fileIndexAfterFilterSafelyDeleteIndex < filesShouldDelete.length) {
+        StringBuilder summary =
+            new StringBuilder(
+                String.format(
+                    "wal node-%s delete outdated summary:the files that should be deleted is: [ %s ],successfully deleted file is: [ %s ],firstValidVersionId is %s .The following reasons influenced the result: %s",
+                    identifier,
+                    Arrays.stream(filesShouldDelete)
+                        .map(File::getName)
+                        .collect(Collectors.joining(",")),
+                    StringUtils.join(successfullyDeleted, ","),
+                    firstValidVersionId,
+                    System.getProperty("line.separator")));
+
+        if (!pinnedMemTableIds.isEmpty()) {
+          summary
+              .append("- MemTable has been flushed but pinned by PIPE,the MemTableId list is : ")
+              .append(StringUtils.join(pinnedMemTableIds, ","))
+              .append(".")
+              .append(System.getProperty("line.separator"));
+        }
+        if (fileIndexAfterFilterSafelyDeleteIndex < filesShouldDelete.length) {
+          summary.append(
+              String.format(
+                  "- MemTable contained in the wal file was flushed, but the data in the file was not consumed by the consensus group.current searchIndex is %d,safelyDeleteIndex is %d",
+                  getCurrentSearchIndex(), safelyDeletedSearchIndex));
+        }
+        String summaryLog = summary.toString();
+        logger.warn(summaryLog);
```

Comment:
Your summary log is very long because of the filesShouldDelete, why use warn log level here?